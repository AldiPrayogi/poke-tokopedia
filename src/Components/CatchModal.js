/** @jsxImportSource @emotion/react */

import React from 'react';
import {Button, Modal, WhiteSpace} from 'antd-mobile';
import 'antd/es/spin/style/css';
import {NameForm} from './NameForm';
import {CatchModalCSS} from "../Styling/PagesCSS";
import {Loading} from "./Loading";

export const CatchModal =  ({visible, setVisible, pokemon, catchChance, setChance, isLoading, setIsLoading}) => {
    const handleOnClose = () => {
        setVisible(false);
        setIsLoading(true);
    }

    const handleChance = (chance) => {
        return chance < 100;
    }

    const handleThrow = () => {
        setTimeout(() => {
            setIsLoading(false);
        }, 2000);
    }

    const handleRetry = () => {
        setIsLoading(true);
        setChance(Math.floor((Math.random() * 100) + 1));
        handleThrow();
    }
    return (
        <div>
            <Modal
                visible={visible}
                markClosable={true}
                transparent
                afterClose={() => handleOnClose()}
                css={CatchModalCSS}
            >
                {isLoading ?
                    <div>
                        <Loading/>
                    </div>
                    :
                    <div>
                        {handleChance(catchChance) ?
                            <div className='successful-catch-container'>
                                <div>
                                    <h3>{pokemon.name}</h3>
                                </div>
                                <div>
                                    <img src={pokemon.sprites.front_default} className='pokemon-image' alt={pokemon.name}/>
                                </div>
                                <NameForm setVisible={setVisible} pokemon={pokemon}/>
                                <WhiteSpace size='xl'/>
                                <WhiteSpace size='xl'/>
                            </div>
                            :
                            <div className='failed-catch-container'>
                                <h1>OH NO, THE POKEMON BROKE FREE!</h1>
                                <WhiteSpace size='md'/>
                                <div className='button-container'>
                                    <Button onClick={handleRetry}>Retry</Button>
                                    <Button onClick={handleOnClose}>Close</Button>
                                </div>
                            </div>
                        }
                    </div>
                }
            </Modal>
        </div>
    );
}