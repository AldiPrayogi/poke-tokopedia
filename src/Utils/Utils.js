import {
    BUG, DARK, DEFAULT, DRAGON, ELECTRIC, FAIRY,
    FIGHTING, FIRE, FLYING, GHOST, GRASS, GROUND, ICE,
    NORMAL, POISON, PSYCHIC, ROCK, STEEL, WATER
} from '../Constants/Color';
import {css} from "@emotion/react";

export const checkType = (type) => {
    switch (type){
        case 'normal':
            return NORMAL
        case 'fighting':
            return FIGHTING
        case 'flying':
            return FLYING
        case 'poison':
            return POISON
        case 'ground':
            return GROUND
        case 'rock':
            return ROCK
        case 'bug':
            return BUG
        case 'ghost':
            return GHOST
        case 'steel':
            return STEEL
        case 'fire':
            return FIRE
        case 'water':
            return WATER
        case 'grass':
            return GRASS
        case 'electric':
            return ELECTRIC
        case 'psychic':
            return PSYCHIC
        case 'dragon':
            return DRAGON
        case 'dark':
            return DARK
        case 'fairy':
            return FAIRY
        case 'ice':
            return ICE
        default:
            return DEFAULT
    }
}

export const checkBackgroundColor = (type) => {
    const backgroundColor = checkType(type)

    return css`
        background-color: ${backgroundColor};
        width: 25vw;
        color: black;
        max-width: 120px;
        padding: 0;
        text-align: center;
        margin-right: 1vw;
        display: inline-block;
        font-size: 100%;
        font-weight: bold;
        border-style: none;
    `
}

export const checkBackgroundColorPokedex = (type) => {
    const backgroundColor = checkType(type)

    return css`
        background-color: ${backgroundColor};
        width: 40%;
        margin-left: 5%;
        margin-top: 5%;
        color: black;
        padding: 0;
        text-align: center;
        margin-right: 5%;
        display: inline-block;
        font-size: 1vh;
        font-weight: bold;
        border-style: none;
    `
}

export const FIRST_TEXT = 'YOU THROW A POKEBALL....';