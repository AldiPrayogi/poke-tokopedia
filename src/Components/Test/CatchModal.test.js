import {shallow} from 'enzyme';
import {CatchModal} from "../CatchModal";
import {DetailedPokemonData} from '../../Fixtures/PokemonData';
import {FIRST_TEXT} from "../../Utils/Utils";

afterEach(() =>{
    jest.clearAllMocks();
})

afterAll(() => {
    jest.clearAllMocks();
})

describe('CatchModal', () => {
    describe('CatchModal Failed Catch', () => {
        const setVisibleMock = jest.fn();
        const setLoadingMock = jest.fn();
        const setLoadingTextMock = jest.fn();
        const setChanceMock = jest.fn();
        const handleLoadingMock = jest.fn();

        const wrapperFailed = shallow(
            <CatchModal
                visible={true}
                setVisible={setVisibleMock}
                setLoadingText={setLoadingTextMock}
                setIsLoading={setLoadingMock}
                handleLoading={handleLoadingMock}
                pokemon={DetailedPokemonData}
                catchChance={60}
                setChance={setChanceMock}
                isLoading={false}
            />
        )

        it('should render correctly with loading false and failed catch', () => {
            expect(wrapperFailed).toMatchSnapshot();
        });

        describe('handleOnClose', () => {
            it('should call setVisible, setLoading and setLoadingTextMock when close button is clicked', () =>{
                const closeButton = wrapperFailed.find('.cancel-button');

                closeButton.simulate('click');

                expect(setVisibleMock).toHaveBeenCalledWith(false);
                expect(setLoadingMock).toHaveBeenCalledWith(true);
                expect(setLoadingTextMock).toHaveBeenCalledWith(FIRST_TEXT);
            })
        });

        describe('handleRetry', () => {
            it('should call setLoadingText, setLoading, setChance, and handleThrow when retry button is clicked', () =>{
                const retryButton = wrapperFailed.find('.retry-button');

                retryButton.simulate('click');

                expect(setLoadingMock).toHaveBeenCalledWith(true);
                expect(setLoadingTextMock).toHaveBeenCalledWith(FIRST_TEXT);
                expect(setChanceMock).toHaveBeenCalled();
                expect(handleLoadingMock).toHaveBeenCalled();
            })
        });
    });

    describe('CatchModal Successful Catch', () => {
        const wrapperSuccess = shallow(
            <CatchModal
                visible={true}
                pokemon={DetailedPokemonData}
                catchChance={40}
                isLoading={false}
            />
        )

        it('should render correctly with loading false and successful catch', () => {
            expect(wrapperSuccess).toMatchSnapshot();
        });
    });

    describe('CatchModal Loading', () => {
        const wrapperSuccess = shallow(
            <CatchModal
                visible={true}
                pokemon={DetailedPokemonData}
                catchChance={40}
                isLoading={true}
                loadingText={FIRST_TEXT}
            />
        )

        it('should render correctly with loading is true', () => {
            expect(wrapperSuccess).toMatchSnapshot();
        });
    });
});