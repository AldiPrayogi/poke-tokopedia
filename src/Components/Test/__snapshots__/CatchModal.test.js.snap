// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatchModal CatchModal Failed Catch should render correctly with loading false and failed catch 1`] = `
<div>
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="CatchModal"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__={[Function]}
    css={
      Object {
        "map": undefined,
        "name": "rxonfe",
        "next": undefined,
        "styles": "
    max-width: 800px;
    margin-bottom: 100px;
    .am-modal-transparent{
        width: 20vh;
    }
    .am-modal-header:{
        font-weight: bold;
        border-radius: 10px;
    }
    h2{
        background-color: #222224;
        color: #f0c44d;
        font-weight: bolder;
    }
    .loading-catch-text{
        h1{
            background-color: #222224;
            color: #f0c44d;
            font-weight: bolder;
        }
        background-color: #373F51;
        padding: 2vh 0 5px 0;
    }
    .pokemon-image{
        width: 50vw;
        max-width: 300px;
        background-image: url('/images/2gfwG.jpg');
    }
    .am-modal-content: {
        background-color: #373F51;
        padding-top: 0px !important;
    }
    .successful-catch-container{
        background-color: #373F51;
        h3{
            font-size: 2.6vw;
            font-weight: bold;
            background-color: #373F51;
            color: white;
        }
        h2{
            font-size: 2.8vh;
        }
    }
    .container{
        text-align: center;
        width: 100%;
        display: inline-block;
    }
    .am-button{
        margin: auto;
        width: 40%;
        display: inline-block;
        color: #f0c44d;
        background-color: #222224;
        border-color: #f0c44d;
    }
    .failed-catch-container{
        h1{
            color: #f0c44d;
            font-weight: bolder;
        }
        background-color: #373F51;
        padding: 0 0 10px 0;
    }
    .failed-catch-text{
        background-color: #222224;
        width: 100%;
        padding: 10px;
    }
    .button-container{
        display: flex;
        justify-content: space-between;
        font-weight: bold;
    }
    width: 80vw;
    text-align: center;
    .button-retry{
        color: #f0c44d;
        background-color: #222224;
    }
",
        "toString": [Function],
      }
    }
    markClosable={true}
    transparent={true}
    visible={true}
  >
    <div>
      <div
        className="failed-catch-container"
      >
        <div
          className="failed-catch-text"
        >
          <h1>
            OH NO, THE POKEMON BROKE FREE!
          </h1>
        </div>
        <WhiteSpace
          prefixCls="am-whitespace"
          size="md"
        />
        <div
          className="button-container"
        >
          <Button
            activeStyle={Object {}}
            className="retry-button"
            disabled={false}
            inline={false}
            loading={false}
            onClick={[Function]}
            prefixCls="am-button"
            size="large"
          >
            Retry
          </Button>
          <Button
            activeStyle={Object {}}
            className="cancel-button"
            disabled={false}
            inline={false}
            loading={false}
            onClick={[Function]}
            prefixCls="am-button"
            size="large"
          >
            Close
          </Button>
        </div>
      </div>
    </div>
  </EmotionCssPropInternal>
</div>
`;

exports[`CatchModal CatchModal Loading should render correctly with loading is true 1`] = `
<div>
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="CatchModal"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__={[Function]}
    css={
      Object {
        "map": undefined,
        "name": "rxonfe",
        "next": undefined,
        "styles": "
    max-width: 800px;
    margin-bottom: 100px;
    .am-modal-transparent{
        width: 20vh;
    }
    .am-modal-header:{
        font-weight: bold;
        border-radius: 10px;
    }
    h2{
        background-color: #222224;
        color: #f0c44d;
        font-weight: bolder;
    }
    .loading-catch-text{
        h1{
            background-color: #222224;
            color: #f0c44d;
            font-weight: bolder;
        }
        background-color: #373F51;
        padding: 2vh 0 5px 0;
    }
    .pokemon-image{
        width: 50vw;
        max-width: 300px;
        background-image: url('/images/2gfwG.jpg');
    }
    .am-modal-content: {
        background-color: #373F51;
        padding-top: 0px !important;
    }
    .successful-catch-container{
        background-color: #373F51;
        h3{
            font-size: 2.6vw;
            font-weight: bold;
            background-color: #373F51;
            color: white;
        }
        h2{
            font-size: 2.8vh;
        }
    }
    .container{
        text-align: center;
        width: 100%;
        display: inline-block;
    }
    .am-button{
        margin: auto;
        width: 40%;
        display: inline-block;
        color: #f0c44d;
        background-color: #222224;
        border-color: #f0c44d;
    }
    .failed-catch-container{
        h1{
            color: #f0c44d;
            font-weight: bolder;
        }
        background-color: #373F51;
        padding: 0 0 10px 0;
    }
    .failed-catch-text{
        background-color: #222224;
        width: 100%;
        padding: 10px;
    }
    .button-container{
        display: flex;
        justify-content: space-between;
        font-weight: bold;
    }
    width: 80vw;
    text-align: center;
    .button-retry{
        color: #f0c44d;
        background-color: #222224;
    }
",
        "toString": [Function],
      }
    }
    markClosable={true}
    transparent={true}
    visible={true}
  >
    <div>
      <div
        className="loading-catch-text"
      >
        <WhiteSpace
          prefixCls="am-whitespace"
          size="sm"
        />
        <h1>
          YOU THROW A POKEBALL....
        </h1>
      </div>
    </div>
  </EmotionCssPropInternal>
</div>
`;

exports[`CatchModal CatchModal Successful Catch should render correctly with loading false and successful catch 1`] = `
<div>
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="CatchModal"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__={[Function]}
    css={
      Object {
        "map": undefined,
        "name": "rxonfe",
        "next": undefined,
        "styles": "
    max-width: 800px;
    margin-bottom: 100px;
    .am-modal-transparent{
        width: 20vh;
    }
    .am-modal-header:{
        font-weight: bold;
        border-radius: 10px;
    }
    h2{
        background-color: #222224;
        color: #f0c44d;
        font-weight: bolder;
    }
    .loading-catch-text{
        h1{
            background-color: #222224;
            color: #f0c44d;
            font-weight: bolder;
        }
        background-color: #373F51;
        padding: 2vh 0 5px 0;
    }
    .pokemon-image{
        width: 50vw;
        max-width: 300px;
        background-image: url('/images/2gfwG.jpg');
    }
    .am-modal-content: {
        background-color: #373F51;
        padding-top: 0px !important;
    }
    .successful-catch-container{
        background-color: #373F51;
        h3{
            font-size: 2.6vw;
            font-weight: bold;
            background-color: #373F51;
            color: white;
        }
        h2{
            font-size: 2.8vh;
        }
    }
    .container{
        text-align: center;
        width: 100%;
        display: inline-block;
    }
    .am-button{
        margin: auto;
        width: 40%;
        display: inline-block;
        color: #f0c44d;
        background-color: #222224;
        border-color: #f0c44d;
    }
    .failed-catch-container{
        h1{
            color: #f0c44d;
            font-weight: bolder;
        }
        background-color: #373F51;
        padding: 0 0 10px 0;
    }
    .failed-catch-text{
        background-color: #222224;
        width: 100%;
        padding: 10px;
    }
    .button-container{
        display: flex;
        justify-content: space-between;
        font-weight: bold;
    }
    width: 80vw;
    text-align: center;
    .button-retry{
        color: #f0c44d;
        background-color: #222224;
    }
",
        "toString": [Function],
      }
    }
    markClosable={true}
    transparent={true}
    visible={true}
  >
    <div>
      <div
        className="successful-catch-container"
      >
        <div>
          <h2>
            CONGRATULATIONS!
          </h2>
        </div>
        <div>
          <img
            alt="Pokemon"
            className="pokemon-image"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png"
          />
        </div>
        <WhiteSpace
          prefixCls="am-whitespace"
          size="md"
        />
        <div>
          <h3>
            GOTCHA!
          </h3>
          <h3>
            POKEMON
             WAS CAUGHT!
          </h3>
        </div>
        <WhiteSpace
          prefixCls="am-whitespace"
          size="md"
        />
        <NameForm
          pokemon={
            Object {
              "abilities": Array [
                Object {
                  "ability": Object {
                    "name": "abc",
                  },
                },
              ],
              "gender": "Male",
              "moves": Array [
                Object {
                  "move": Object {
                    "name": "abc",
                  },
                },
              ],
              "name": "Pokemon",
              "nickName": "Bulb",
              "pokemon": Object {
                "name": "pokemon",
              },
              "sprites": Object {
                "front_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png",
              },
              "types": Array [
                Object {
                  "type": Object {
                    "name": "abc",
                  },
                },
              ],
            }
          }
        />
      </div>
    </div>
  </EmotionCssPropInternal>
</div>
`;
